rsocket-micrometer/src/main/java/io/rsocket/micrometer/MicrometerDuplexConnection.java

```diff
@@ -0,0 +1,244 @@
+import static io.rsocket.framing.FrameType.KEEPALIVE;
+import static io.rsocket.framing.FrameType.LEASE;
+import static io.rsocket.framing.FrameType.METADATA_PUSH;
+import static io.rsocket.framing.FrameType.NEXT;
+import static io.rsocket.framing.FrameType.NEXT_COMPLETE;
+import static io.rsocket.framing.FrameType.REQUEST_CHANNEL;
+import static io.rsocket.framing.FrameType.REQUEST_FNF;
+import static io.rsocket.framing.FrameType.REQUEST_N;
+import static io.rsocket.framing.FrameType.REQUEST_RESPONSE;
+import static io.rsocket.framing.FrameType.REQUEST_STREAM;
+import static io.rsocket.framing.FrameType.RESUME;
+import static io.rsocket.framing.FrameType.RESUME_OK;
+import static io.rsocket.framing.FrameType.SETUP;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+import io.rsocket.DuplexConnection;
+import io.rsocket.Frame;
+import io.rsocket.framing.FrameType;
+import io.rsocket.plugins.DuplexConnectionInterceptor.Type;
+import java.util.Objects;
+import java.util.function.Consumer;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+/**
+ * An implementation of {@link DuplexConnection} that intercepts frames and gathers Micrometer
+ * metrics about them.
+ *
+ * <p>The metric is called {@code rsocket.frame} and is tagged with {@code connection.type} ({@link
+ * Type}), {@code frame.type} ({@link FrameType}), and any additional configured tags.
+ *
+ * @see <a href="https://micrometer.io">Micrometer</a>
+ */
+final class MicrometerDuplexConnection implements DuplexConnection {
+
+  private final DuplexConnection delegate;
+
+  private final FrameCounters frameCounters;
+
+  /**
+   * Creates a new {@link DuplexConnection}.
+   *
+   * @param connectionType the type of connection being monitored
+   * @param delegate the {@link DuplexConnection} to delegate to
+   * @param meterRegistry the {@link MeterRegistry} to use
+   * @param tags additional tags to attach to {@link Meter}s
+   * @throws NullPointerException if {@code connectionType}, {@code delegate}, or {@code
+   *     meterRegistry} is {@code null}
+   */
+  MicrometerDuplexConnection(
+      Type connectionType, DuplexConnection delegate, MeterRegistry meterRegistry, Tag... tags) {
+
+    Objects.requireNonNull(connectionType, "connectionType must not be null");
+    this.delegate = Objects.requireNonNull(delegate, "delegate must not be null");
+    Objects.requireNonNull(meterRegistry, "meterRegistry must not be null");
+
+    this.frameCounters = new FrameCounters(connectionType, meterRegistry, tags);
+  }
+
+  @Override
+  public void dispose() {
+    delegate.dispose();
+  }
+
+  @Override
+  public Mono<Void> onClose() {
+    return delegate.onClose();
+  }
+
```

Comment:
What are people's thoughts on counting the disposes and onCloses? Might help with debugging or alerts

Reply:
I don't really have enough historical context to know what problem that's trying to solve.  I'll defer to the rest of you for a definitive call on collecting them.  Just let me know and I'll add it if desired.

Reply:
I've added onClose print statements to debug  issues and I've had similar data about http before so I would like to capture it if no one has an objection.

Reply:
Will do @robertroeser.  My suspicion is that this will only be 0/1.  Is that correct?